src/big_int10.cpp

```
@@ -0,0 +1,562 @@
+    remainder.limbs.clear();
+    std::vector<uint8_t> quotient;
+
+    for (int i = a.limbsCount() - 1; i >= 0; i--) {  // msb-lsb
+
+        remainder.limbs.insert(remainder.limbs.begin(),
+                               a.limbs[i]);  // insert msb as lsb
+        if (remainder.limbsCount() > 1 && remainder.limbs.back() == 0)
+            remainder.limbs.pop_back();  // remove 1 leading zero
+
+        int times = 0;
+        if (!remainder.isSmallerThanUnsigned(b)) {
+            int left = 1, right = 9;
+            while (left <= right) {
+                int mid = left + (right - left) / 2;
+                if (!remainder.isSmallerThanUnsigned(
+                        b * mid)) {  // b*times<=remainder
+                    times = mid;
+                    if (left == right)
+                        break;
+                    left = mid + 1;
+                }
+                else {
+                    right = mid - 1;
+                }
+            }
+            remainder.subtractUnsigned(b * times);
+        }
+
+        if (quotient.size() > 0 || times > 0)
+            quotient.push_back(times);
+    }
+
+    std::reverse(quotient.begin(), quotient.end());
+    BigInt10 quoti(a.isNegative != b.isNegative);
+    quoti.limbs = std::move(quotient);
+    return quoti;
+}
+BigInt10 power(BigInt10 base, BigInt10 exponent)
+{
+    if (exponent.isNegative)
+        throw std::invalid_argument("Invalid input: negative exponent");
+
+    BigInt10 result(1);
+
+    while (!exponent.isZero()) {
+        if (exponent.limbs[0] & 1) {
+            result *= base;
+        }
+        base *= base;
+        exponent /= 2;
+    }
+    return result;
+}
+void BigInt10::initFromString(const char *str, int n)
+{
+    isNegative = str[0] == '-';
+
+    limbs.reserve(n);
+    for (int i = n - 1; i >= isNegative; i--) {
+        uint8_t digit = str[i] - '0';
+        if (!isDigit10(digit))
+            throw std::invalid_argument("Invalid input: non-digit character");
```

Should return an error code?

---

wrapped with try catch